framework/Util/TLogger.php

```diff
@@ -33,49 +43,388 @@ class TLogger extends \Prado\TComponent
 	public const ERROR = 0x10;
 	public const ALERT = 0x20;
 	public const FATAL = 0x40;
+
+	public const PROFILE = 0x100;
+	public const PROFILE_BEGIN = 0x300;
+	public const PROFILE_END = 0x500;
+
+	/**
+	 * These are for selecting just the PROFILE_BEGIN and PROFILE_END logs without
+	 * the PROFILE bit.
+	 */
+	public const PROFILE_BEGIN_SELECT = 0x200;
+	public const PROFILE_END_SELECT = 0x400;
+
+	public const LOGGED = 0x8000;
+
+	/**
+	 * Log Information Index.
+	 */
+	public const LOG_MESSAGE = 0;
+	public const LOG_LEVEL = 1;
+	public const LOG_CATEGORY = 2;
+	public const LOG_TIME = 3;
+	public const LOG_MEMORY = 4;
+	public const LOG_CONTROL = 5;
+	public const LOG_TRACES = 6;
+	public const LOG_PID = 7;
+
 	/**
 	 * @var array log messages
 	 */
-	private $_logs = [];
+	private array $_logs = [];
 	/**
-	 * @var int log levels (bits) to be filtered
+	 * @var array unmatched PROFILE_BEGIN log messages
+	 * @since 4.2.3
 	 */
-	private $_levels;
+	private array $_profileLogs = [];
 	/**
-	 * @var array list of categories to be filtered
+	 * @var int the profileLogs count that are not static::LOGGED.
+	 * @since 4.2.3
 	 */
-	private $_categories;
+	private int $_profileLogsCount = 0;
+	/**
+	 * @var array The maintained Profile Begin times
+	 * @since 4.2.3
+	 */
+	private array $_profileBeginTimes = [];
+	/**
+	 * @var ?int log levels (bits) to be filtered
+	 */
+	private ?int $_levels = null;
+	/**
+	 * @var ?array list of categories to be filtered
+	 */
+	private ?array $_categories = null;
 	/**
 	 * @var array list of control client ids to be filtered
 	 */
-	private $_controls;
+	private ?array $_controls = null;
+	/**
+	 * @var ?float timestamp used to filter
+	 */
+	private ?float $_timestamp;
+	/**
+	 * @var ?int process id used to filter
+	 * @since 4.2.3
+	 */
+	private ?int $_pid;
+
+	/**
+	 * @var int the number of logs before flushing.
+	 * @since 4.2.3
+	 */
+	private int $_flushCount = 1000;
+	/**
+	 * @var bool is the logger flushing.
+	 * @since 4.2.3
+	 */
+	private bool $_flushing = false;
+	/**
+	 * @var ?array any logged messages during flushing so they aren't flushed.
+	 * @since 4.2.3
+	 */
+	private ?array $_flushingLog = null;
+	/**
+	 * @var int The depth of a trace, default 0 for no trace.
+	 * @since 4.2.3
+	 */
+	private int $_traceLevel = 0;
+	/**
+	 * @var bool Is the logger Registered with onEndRequest
+	 * @since 4.2.3
+	 */
+	private bool $_registered = false;
+
+	/**
+	 * @param bool $flushShutdown Should onFlushLogs be a register_shutdown_function.
+	 * @since 4.2.3
+	 */
+	public function __construct(bool $flushShutdown = true)
+	{
+		if($flushShutdown) {
+			register_shutdown_function(function () {
+				$this->onFlushLogs();
+				register_shutdown_function([$this, 'onFlushLogs'], $this, true);
```

Comment:
This looks strange, a register_shutdown_function() inside a register_shutdown_function()

Reply:
Yii2 and other frameworks are models.   The second shutdown is to wrap up any logs entered by shutdown functions.   The second time is appended to the list of shutdown functions.

Reply:
Nice, i didn't think of that.

Reply:
There are a lot of great ideas I pulled from other projects like this.   I also found a lot of "holes" that other (like Yii2) were missing or doing in a much more complex way than needed.     eg: capturing the pid.  I have a some solid child forking code coming up that will actually capture a forked child log, send it back to the main thread and log it there.  So knowing the pid of the log will become important.